signer/src/network/libp2p/swarm.rs

```diff
@@ -0,0 +1,273 @@
+                gossipsub_config,
+            )
+            .map_err(SignerSwarmError::LibP2PMessage)?,
+            mdns: mdns::tokio::Behaviour::new(
+                mdns::Config::default(),
+                keypair.public().to_peer_id(),
+            )
+            .map_err(|e| SignerSwarmError::LibP2P(Box::new(e)))?,
+            kademlia: kad::Behaviour::new(
+                keypair.public().to_peer_id(),
+                MemoryStore::new(keypair.public().to_peer_id()),
+            ),
+            ping: ping::Behaviour::default(),
+            relay: relay::Behaviour::new(keypair.public().to_peer_id(), Default::default()),
+            identify: identify::Behaviour::new(identify::Config::new(
+                "/sbtc-signer/1.0.0".into(),
+                keypair.public(),
+            )),
+            autonat_server: AutoNatServerBehavior::new(OsRng),
+            autonat_client: AutoNatClientBehavior::new(
+                OsRng,
+                AutoNatClientConfig::default().with_probe_interval(Duration::from_secs(2)),
+            ),
+        })
+    }
+}
+
+/// Builder for the [`SignerSwarm`] libp2p network.
+pub struct SignerSwarmBuilder<'a> {
+    private_key: &'a PrivateKey,
+    listen_on: Vec<Multiaddr>,
+    seed_addrs: Vec<Multiaddr>,
+}
+
+impl<'a> SignerSwarmBuilder<'a> {
+    /// Create a new [`SignerSwarmBuilder`] with the given private key.
+    pub fn new(private_key: &'a PrivateKey) -> Self {
+        Self {
+            private_key,
+            listen_on: Vec::new(),
+            seed_addrs: Vec::new(),
+        }
+    }
+
+    /// Add a listen endpoint to the builder.
+    #[allow(dead_code)]
+    pub fn add_listen_endpoint(mut self, addr: Multiaddr) -> Self {
+        if !self.listen_on.contains(&addr) {
+            self.listen_on.push(addr);
+        }
+        self
+    }
+
+    /// Add multiple listen endpoints to the builder.
+    #[allow(dead_code)]
+    pub fn add_listen_endpoints(mut self, addrs: &[Multiaddr]) -> Self {
+        for addr in addrs {
+            if !self.listen_on.contains(addr) {
+                self.listen_on.push(addr.clone());
+            }
+        }
+        self
+    }
+
+    /// Add a seed address to the builder.
+    #[allow(dead_code)]
+    pub fn add_seed_addr(mut self, addr: Multiaddr) -> Self {
+        if !self.seed_addrs.contains(&addr) {
+            self.seed_addrs.push(addr);
+        }
+        self
+    }
+
+    /// Add multiple seed addresses to the builder.
+    #[allow(dead_code)]
+    pub fn add_seed_addrs(mut self, addrs: &[Multiaddr]) -> Self {
+        for addr in addrs {
+            if !self.seed_addrs.contains(addr) {
+                self.seed_addrs.push(addr.clone());
+            }
+        }
+        self
+    }
+
+    /// Build the [`SignerSwarm`], consuming the builder.
+    pub fn build(self) -> Result<SignerSwarm, SignerSwarmError> {
+        let keypair = Keypair::ed25519_from_bytes(self.private_key.to_bytes())?;
```

Comment:
Curious, why the `ed25519` curve instead of `secp256k1`?

Reply:
Not by choice, it's the only method I found to get a `Keypair` from bytes. Seems that libp2p only supports ed25519. Do you think this will be an issue?

Reply:
I don't think it will be an issue. The messages are signed with ECDSA signatures over `secp256ka` and they are transported over a p2p network which, while using the same keys, uses EdDSA signatures over Curve25519. Those doesn't have to have anything to do with each other and the fact that we're reusing the same keys is more of a convenience than a necessity here.

Reply:
Referencing @netrome's comment below:
https://github.com/stacks-network/sbtc/pull/473#discussion_r1745996675

Reply:
Yeah, I didn't think this will be an issue either. But I believe `secp256k1` is supported via a feature flag. The thought that I had is that we can easily create a `secp256k1` variant of the `libp2p::Keypair` from our internal types, and I was curious if there was any reason why we should not do that.

Reply:
> The thought that I had is that we can easily create a secp256k1 variant of the libp2p::Keypair from our internal types

This sounds nicer, i'll look into it 😊

Reply:
We can do that after this PR too. I just wanted to share what I was thinking.